### PR TITLE
Implement improved event dispatching

### PR DESCRIPTION
**Pull request by jordangarcia**
## Summary
- Introduce the `js-sdk-event-processor` package
- Send Impression and Conversion events to the event processor where batching can happen.

## Test plan
- Unit tests

## Issues
- OASIS-4123

### CODE REVIEW

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 1]
```diff
@@ -102,8 +110,17 @@ function Optimizely(config) {
...
+    errorHandler: this.errorHandler,
+  });
+
+  this.eventProcessor = new eventProcessor.LogTierV1EventProcessor({
+    dispatcher: this.eventDispatcher,
+    flushInterval: config.eventFlushInterval || DEFAULT_EVENT_FLUSH_INTERVAL,
+    maxQueueSize: config.eventBatchSize || DEFAULT_EVENT_MAX_QUEUE_SIZE,
   });
+  // TODO figure out some way to specify batching vs non-batching
+  // to help be backwards compatible with existing implementation
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Can you expand on this, and why it's left as TODO for now?

Don't `config.eventBatchSize` and `config.eventFlushInterval` defaults let us achieve backwards compatibility?

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 2]
```diff
@@ -42,6 +42,13 @@ var LOG_MESSAGES = enums.LOG_MESSAGES;
 var DECISION_SOURCES = enums.DECISION_SOURCES;
 
 describe('lib/optimizely', function() {
+
+  afterEach(function(done) {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: This seems like a last resort when we're in async test hell and don't know what's causing problems.

File: packages/optimizely-sdk/lib/optimizely/index.tests.js [hunk 3]
```diff
@@ -872,7 +752,7 @@ describe('lib/optimizely', function() {
           datafile: testData.getTestProjectConfig(),
           errorHandler: errorHandler,
           eventDispatcher: eventDispatcher,
-          jsonSchemaValidator: jsonSchemaValidator,
+          skipJSONValidation: true,
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/index.tests.js]: Why this change? Is it related to event-processor?

File: packages/optimizely-sdk/lib/optimizely/event_dispatcher_bridge.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      url: request.url,
+      params: request.event,
+    },
+    function(response) {
+      // right now callbacks only happen if statusCode >= 200 && < 400
+      callback(true);
+    }
+  );
+
+  if (typeof maybePromise === 'object' && typeof maybePromise.then === 'function') {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/optimizely/event_dispatcher_bridge.js]: Need to check maybePromise is not `null` before accessing `maybePromise.then`?

Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/event_dispatcher_bridge.js]: good catch

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 1 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.js]: Oh this comment is outdated, this is handled by using `eventBatchSize=1`

File: packages/optimizely-sdk/lib/optimizely/index.tests.js (hunk 3 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js]: fixed

File: packages/optimizely-sdk/lib/optimizely/index.tests.js (hunk 2 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.tests.js]: not needed, removed
